cpp/daal/src/algorithms/tsne/tsne_gradient_descent_avx512_impl.i

```diff
@@ -0,0 +1,326 @@
+//--
+*/
+
+#ifndef __TSNE_GRADIENT_DESCENT_AVX512_IMPL_I__
+#define __TSNE_GRADIENT_DESCENT_AVX512_IMPL_I__
+
+namespace daal
+{
+namespace algorithms
+{
+namespace internal
+{
+
+/* Partial template specialization of attractive kernel for single precision data and AVX512 ISA */
+template <bool DivComp, typename IdxType>
+struct AttractiveKernel<DivComp, IdxType, float, avx512>
+{
+    static services::Status impl(const float * val, const IdxType * col, const size_t * row, MemoryCtxType<IdxType, float, avx512> & mem,
+                                 float & zNorm, float & divergence, const IdxType N, const IdxType nnz, const IdxType nElements,
+                                 const float exaggeration)
+    {
+        const float multiplier = exaggeration * float(zNorm);
+        divergence             = 0.;
+
+        const IdxType prefetch_dist = 32;
+
+        daal::TlsSum<float, avx512> divTlsData(1);
+
+        SafeStatus safeStat;
+        daal::tls<float *> logTlsData([=, &safeStat]() {
+            auto logData = services::internal::service_scalable_calloc<float, avx512>(nElements);
+            if (!logData)
+            {
+                safeStat.add(services::ErrorMemoryAllocationFailed);
+            }
+            return logData;
+        });
+
+        const IdxType nThreads    = threader_get_threads_number();
+        const IdxType sizeOfBlock = services::internal::min<avx512, size_t>(256, N / nThreads + 1);
+        const IdxType nBlocks     = N / sizeOfBlock + !!(N % sizeOfBlock);
+
+        daal::threader_for(nBlocks, nBlocks, [&](IdxType iBlock) {
+            const IdxType iStart = iBlock * sizeOfBlock;
+            const IdxType iEnd   = services::internal::min<avx512, IdxType>(N, iStart + sizeOfBlock);
+            float * logLocal     = logTlsData.local();
+            if (!logLocal) return;
+
+            float * divLocal = divTlsData.local();
+            DAAL_CHECK_MALLOC_THR(divLocal);
+
+            xyType<float> row_point;
+            IdxType iCol, prefetch_index;
+            float y1d, y2d, sqDist, PQ;
+
+            for (IdxType iRow = iStart; iRow < iEnd; ++iRow)
+            {
+                size_t iSize      = 0;
+                mem._attr[iRow].x = 0.0;
+                mem._attr[iRow].y = 0.0;
+                row_point         = mem._pos[iRow];
+
+                if (!DivComp)
+                {
+                    IdxType start_index = row[iRow] - 1;
+                    IdxType range       = row[iRow + 1] - row[iRow];
+
+                    __m512 vec_1   = _mm512_set1_ps(1.0);
+                    __m512i vec_1i = _mm512_set1_epi32(1);
+
+                    __m512 vec_point_x = _mm512_set1_ps(row_point.x);
+                    __m512 vec_point_y = _mm512_set1_ps(row_point.y);
+
+                    __m512 vec_point_xs = _mm512_setzero_ps();
+                    __m512 vec_point_ys = _mm512_setzero_ps();
+
+                    for (IdxType i = 0; i < (range / 16) * 16; i += 16)
```

Comment:
1. I see problems with the tail part of this cycle. You are running this loop `range / 16` times
2. Please use separate variable. Something like `constexpr IdxType vec_width = 16`.